## Author: Lyndsey Pohl
# Assignment 12: *Classes*
***

## Introduction
In this assignment, you will be learning about Python classes through practice. As you know from the lesson materials, classes and objects are fundamental concepts in object oriented programming. But to understand this powerful programming model, let's consider the model of programming that we have been using so far. 

## Procedural Programming
Thus far, you have been writing programs that execute a series of commands/calculations to achieve some goal. Your scripts consisted of mainly 3 types of Python objects: functions, iterators, and data containers. These are all objects in Python (they are builtin types from the standard library). So if these are objects, is it fair to say that you have learned object-oriented programming? *Using* objects (like functions) and *creating* objects are different interactions with classes in Python. 

Your experience in this class has so far been with coding according to the "procedural programming paradigm". Procedural programming simply refers to the use of procedures (functions or "routines") to execute a series of computational steps. Our focus has been on writing "functions or procedures which operate on data" <sup>[1]</sup>. In writing every program, we were mostly concerned with identifying and handling variables, manipulating data structures, and performing calculations, usually all in sequence. Contrast this to object-oriented programming, where the "focus is on the creation of classes and objects which contain **both data and functionality together**"<sup>[1]</sup>.

## Object-Oriented Programming *(OOP)*

### Definitional understanding

* When you create a class, you are creating a blueprint for an object. The idea is no different from creating a blueprint for a building or a car. 
* The object, then, is an instance of that blueprint (class). Said differently: an object is a representation or occurrence of a class. 

Classes represent the structure of objects. They contain data and methods, which are known as "functions inside of a class". Methods perform some calculations on object/instance attributes. There are other kinds of methods as well (which perform operations on the *class* or on unrelated namespaces). For the purpose of this assignment, think of methods as functions that perform some task with respect to the **object** only.

Every time that you create an object, you are "instantiating" the class. This is to say, you are creating an *instance* of a class. This instance now has specific attributes and methods. 

### Conceptual understanding

Let's explore classes and objects with an example: airplanes. There are many kinds of planes. There are commercial planes, such as the Boeing 777 or Airbus 330. There are cargo planes, such as the Antonov-225 (civilian purposes). There military cargo planes, such as the Boeing Globemaster. There are civilian/corporate planes such as the Gulf Stream G650. And there are military jets, such as the F-22 Raptor. 

You get the idea: all of these are planes. If we were to represent these planes programmatically, in OOP, we would do this with a class. And this class would be called *planes*. Now, think about the characteristics of planes. What do they all usually have in common? Usually, they all have two wings, a vertical stabilizer, horizontal stabilizer, engines, landing gear, etc. These are what we call class *attributes*. More specfically, they are object attributes.

Let's say we are writing a program and we want to "manifest" or create a plane. How would we do this? We would create a plane by instantiating our blueprint. Yes, the *Plane* class! So, by creating a plane from the *Plane* class, we are creating an instance of a class. In the process, we are passing arguments to our class, which are initiated as object attributes. 

### An example

Let's work through a small example. First, create the *Plane* class:

In [1]:
class Plane(object):
    # Notice the 'self' keyword. It refers to the object itself. 
    # Notice the other parameters. These are placeholders for arguments.
    # What arguments, you ask? The arguments that we will pass to this class when we create an object.
	def __init__(self, name, type, capacity, range, airline):
        # They will be used to define the instance variables below.
		self.name = name
		self.type = type
		self.capacity = capacity
		self.range = range
		self.airline = airline

    # This is a method (function inside of a class). It performs some task on the object. 
    # Again, notice the 'self' keyword. It is a reference to the instance.
	def extend_capacity(self, additional_seats):
		self.capacity += additional_seats
		return self.capacity

    # Another method
	def change_owner(self, new_owner):
		self.airline = new_owner
		return self.airline

    # And another method
	def show_details(self):
		print(f"Airline: {self.airline}. Capacity: {self.capacity}. Name: {self.name}. Type: {self.type}")

### Now, let's create an object:

In [2]:
# instantiate object
new_plane = Plane("Boeing 777", "Wide-body Twinjet", 315, 8555, "Singapore Airlines")

# Show details of new plane
new_plane.show_details()

# Say that Singapore Airlines wants to extend the capacity of its airplane by 81 seats.
new_capacity = new_plane.extend_capacity(81)
print(f"The new capacity of Singapore's {new_plane.name} is: {new_capacity}")

Airline: Singapore Airlines. Capacity: 315. Name: Boeing 777. Type: Wide-body Twinjet
The new capacity of Singapore's Boeing 777 is: 396


In [3]:
# Say that Singapore Airlines sells this plane to EVA Airlines.  
new_plane.change_owner("EVA Air")
print(f"The new owner of Singapore's {new_plane.name} is: {new_plane.airline}")

The new owner of Singapore's Boeing 777 is: EVA Air


Notice that we didn't need to store the results of `new_plane.change_owner` in a variable--calling the class method `change_owner` already changes the value of the `airline` attribute 
of the variable `new_plane`.  (The same is also true of `extend_capacity`, above.)

In [4]:
# What if we wanted to print all of the details of this object? Let's call the show_details method:
new_plane.show_details()

Airline: EVA Air. Capacity: 396. Name: Boeing 777. Type: Wide-body Twinjet


### What is *$__init__$* ?

*__init__* is a reserved method in Python. It is preceded and followed by two underscores (called "dunders" in Python). The technical role of *init* is to initiate instance variables (called attributes) after the object is created. The concept of *__init__* is closely related to that of a constructor in object-oriented programming. Though in Python, it is not formally a constructor, because *init* is called after the object is created/loaded into memory. See sources [6] and [7] for more details.


### What is the *self* keyword?

The *self* keyword is a convention used in Python to refer the object itself. It is a reference to the instance of the class, and any method with *self* as an argument will operate on object itself. In Python, you have to explicitly declare this parameter for a method to operate on the instance. See sources [4], [5] and [6] for more details. There are also class methods (which operate on the class only) and static methods (which operate neither on the class or object). 

### Why use classes?

You've seen how classes are used, above and elsewhere, but you may not understand why or when to use them. Let's cover the benefits. The benefits of using classes are <sup>[2][3]</sup>:

* **Organizing data and functions together**. Classes describe both data and the procedures designed to operate on that data. They are a perfect way to store and structure contextually similar data and functions. This benefit is often referred to as encapsulation.
* **Data persistence**. Not only do objects persist in memory for the duration of program execution, but their state can be stored as a binary object, saved on your system. You can load that object into memory in the future and resume operations.
* **Inheritance**. Parent classes (usually with broad/general utility) can be inherited by subclasses for purpose of extending the functionality of the parent class.
* **Reusability**. Easy generation of objects (and data) is an inherentantly attractive feature of classes.

# Assignment 12: *Practice*
***

To build on your current knowledge and understanding of object-oriented programing, we're going to practice creating classes and analyzing data using a [Cars Dataset](https://think.cs.vt.edu/corgis/json/cars/cars.html). It is in JSON format and is similar to the Cars dataset that you might have seen in R. 

## Review dataset metadata

Let's begin by reviewing the metadata of the dataset above (created by Austin Cory Bart, Virginia Tech). Click the link above and review the field descriptions table.

## Review sample JSON

Next, let's review a truncated version of the dataset above. It's simply 1 row of data, so that you can practice subsetting elements from JSON's dictionary-based format.

## Steps 1 and 2: Practice loading and printing JSON files

1) Start by importing the [json](https://docs.python.org/3/library/json.html) package from Python's standard library.

🎯 Read the documentation for `json.loads` and `json.load`.  What is the difference between these functions?  Create a markdown cell below and write your answer.

The s in json.loads stands for 'string'. This means that json.loads deserializes a string that contains a json document, as opposed to json.load that loads a json file.

2) Using Python's [with](https://jeffknupp.com/blog/2016/03/07/improve-your-python-the-with-statement-and-context-managers/) statement, load the 'cars_truncated.json' file (keyword: file, not string) into a Python object.

### Example:

In [5]:
import json

with open('cars_truncated.json', 'r') as file:
    file = json.load(file)
    print(file)

[{'Engine_Information': {'Transmission': '6 Speed Automatic Select Shift', 'Engine_Type': 'Audi 3.2L 6 cylinder 250hp 236ft-lbs', 'Engine_Statistics': {'Horsepower': 250, 'Torque': 236}, 'Hybrid': False, 'Number of Forward Gears': 6, 'Driveline': 'All-wheel drive'}, 'Identification': {'Make': 'Audi', 'Model_Year': '2009 Audi A3', 'ID': '2009 Audi A3 3.2', 'Classification': 'Automatic transmission', 'Year': 2009}, 'Dimensions': {'Width': 202, 'Length': 143, 'Height': 140}, 'Fuel_Information': {'Highway_mpg': 25, 'City_mpg': 18, 'Fuel_Type': 'Gasoline'}}, {'Engine_Information': {'Transmission': '6 Speed Automatic Select Shift', 'Engine_Type': 'Audi 2.0L 4 cylinder 200 hp 207 ft-lbs Turbo', 'Engine_Statistics': {'Horsepower': 200, 'Torque': 207}, 'Hybrid': False, 'Number of Forward Gears': 6, 'Driveline': 'Front-wheel drive'}, 'Identification': {'Make': 'Audi', 'Model_Year': '2009 Audi A3', 'ID': '2009 Audi A3 2.0 T AT', 'Classification': 'Automatic transmission', 'Year': 2009}, 'Dimensio

Notice the output above. It is a list of nested dictionaries. Again, think about how you would access some element inside of this list. Let's say we wanted to know the 'transmission' of the vehicle. How would we pull that value? 

## Step 3: retrieving a specific value from JSON dictionary, demonstrated

In [6]:
transmission = file[0]['Engine_Information']['Transmission']
print(transmission)

6 Speed Automatic Select Shift


So, the [0] indicates that we're taking the 0th element (i.e., car) in the list.  Then we look in the value of the 'Engine_Information' key (which is itself a dictionary).  And inside that, we look in the value of the 'Transmission' key.

If we wanted to know the transmission type of all vehicles in the dataset, all we would need to do is increment the list index to loop through an entire list of dictionaries. From 0 --> 1 ---> 2 ---> 3... and so on.

# Assignment 12: *Test your understanding*
***

Now that you know how to subset a list of dictionaries, let's apply everything that we have learned so far.

## 🎯 Part 1 - Create a class for the Cars dataset

At the bottom of part 1 is a code cell. This is where you write your class. First, read steps 1 and 2, then write your class.

### Step 1 - Initialize the following 2 instance variables: 
Notice, they are prefixed with the keyword *self*.

* **self.all_cars_by_make** - make this a *dictionary*. Recall, dictionaries consist of keys and values. The keys in this dictionary will be car 'makes'. The value of each key will be a single list of multiple dictionaries - with each dictionary containing the fields in step 2 *for every model_id* of that make.
* **self.all_cars_list** - make this a *list*. This will be a list of lists. Each sub-list is for a distinct model_id, and contains the fields in step 2.

In the code cell at the bottom of part 1, we have defined the variables for you. But you must populate them (step 2).

### Step 2 - Populate both containers (from step 1) with values for the following fields:
* Make
* Model_year
* Model_id
* year
* horsepower
* highway_mpg
* city_mpg
* width
* hybrid

Below is an example of how **self.all_cars_by_make** will look like for Acura and Aston Martin after you have populated this dictionary with values (_only first couple models shown for each make_):

{'Acura': [   {   'hybrid': False,
                     'city_mpg': 20,
                     'highway_mpg': 29,
                     'horsepower': 280,
                     'model_id': '2012 Acura TL',
                     'model_year': '2012 Acura TL',
                     'width': 87,
                     'year': 2012},
                 {   'hybrid': False,
                     'city_mpg': 18,
                     'highway_mpg': 26,
                     'horsepower': 305,
                     'model_id': '2012 Acura TL SH-AWD',
                     'model_year': '2012 Acura TL',
                     'width': 87,
                     'year': 2012}],
'Aston Martin': [   {   'hybrid': False,
                            'city_mpg': 11,
                            'highway_mpg': 17,
                            'horsepower': 470,
                            'model_id': '2011 Aston Martin DB9 Volante',
                            'model_year': '2011 Aston Martin DB9 Volante',
                            'width': 83,
                            'year': 2011},
                        {   'hybrid': False,
                            'city_mpg': 13,
                            'highway_mpg': 20,
                            'horsepower': 470,
                            'model_id': '2011 Aston Martin DB9 Volante AT',
                            'model_year': '2011 Aston Martin DB9 Volante',
                            'width': 83,
                            'year': 2011}]}

Below is an example of **self.all_cars_list** (only first couple elements shown):

[   ['2009 Audi A3', '2009 Audi A3 3.2', 250, 2009, 25, 18, 202, False],
    ['2009 Audi A3', '2009 Audi A3 2.0 T AT', 200, 2009, 28, 22, 202, False]]


### 🎯 Add logic for step 2 in the code cell below:

Notice, the class is already partially written to help guide your thinking.

In [7]:
class Cars(object):
    def __init__(self, filename):      
        # read in JSON file
        with open(filename, 'r') as cars_data:
            # create Python object of JSON data called 'cars_list'
            cars_list = json.load(cars_data)
        # create two containers: a dict with car make as key, and a list of all cars by model_idd
        self.all_cars_by_make = {}
        self.all_cars_list = []
        for i in range(len(cars_list)):
            self.all_cars_by_make.setdefault(cars_list[i]['Identification']['Make'],[]).append({'hybrid':cars_list[i]['Engine_Information']['Hybrid'], 'city_mpg':cars_list[i]['Fuel_Information']['City_mpg'], 'highway_mpg':cars_list[i]['Fuel_Information']['Highway_mpg'], 'horsepower':cars_list[i]['Engine_Information']['Engine_Statistics']['Horsepower'],'model_id':cars_list[i]['Identification']['ID'], 'model_year':cars_list[i]['Identification']['Model_Year'], 'width':cars_list[i]['Dimensions']['Width'], 'year':cars_list[i]['Identification']['Year']  })
            self.all_cars_list.append([cars_list[i]['Identification']['Model_Year'],cars_list[i]['Identification']['ID'], cars_list[i]['Engine_Information']['Engine_Statistics']['Horsepower'], cars_list[i]['Identification']['Year'], cars_list[i]['Fuel_Information']['Highway_mpg'],cars_list[i]['Fuel_Information']['City_mpg'], cars_list[i]['Dimensions']['Width'],cars_list[i]['Engine_Information']['Hybrid']   ])
           
  
    def get_all_cars_by_make(self):
        return self.all_cars_by_make;
  
    def get_all_cars_list(self):
        return self.all_cars_list;
  
    def find_model_id(self, make):
        num_ids = len(self.all_cars_by_make[make])
        model_ids = []
        for i in range(num_ids):
            model_ids.append(self.all_cars_by_make[make][i]['model_id'])
          
        return num_ids, model_ids;
  
    def find_max_horsepower(self, model_ids):
        max_horse = 0
        max_model_id = 0
        tie = False
       
        for model_id in model_ids:
            for i in range(len(self.all_cars_list)):
                if model_id == self.all_cars_list[i][1]:
                    if self.all_cars_list[i][2] >= max_horse:
                        if self.all_cars_list[i][2] > max_horse:
                            max_horse = self.all_cars_list[i][2]
                            max_model_id = self.all_cars_list[i][1]
                            tie = False
                            
                        else:
                           
                            if tie:
                                max_model_id.append(self.all_cars_list[i][1])
                            else:
                                max_model_id_tie = []
                                max_model_id_tie.append(max_model_id )
                                max_model_id_tie.append(self.all_cars_list[i][1] )
                                max_model_id = max_model_id_tie
                            
                            tie = True
                           
        return (max_horse, max_model_id);
  
    def find_mpgratio(self, make):
        max_chratio = 0
      
        for i in range(len(self.all_cars_by_make[make])):
            ratio = float(self.all_cars_by_make[make][i]['city_mpg'])/float(self.all_cars_by_make[make][i]['highway_mpg'])
            if ratio >= max_chratio:
                if ratio > max_chratio:
                    max_chratio = ratio
                    max_model_id = self.all_cars_by_make[make][i]['model_id']
                  
                else:
                    max_chratio = ratio
                    max_model_id_tie = []
                    max_model_id_tie.append(max_model_id)
                    max_model_id_tie.append(self.all_cars_by_make[make][i]['model_id'])
                    max_model_id = max_model_id_tie
                   
                    
       
        
        return (max_chratio, max_model_id);
  
    def count_hybrid(self, year):
        count_model_ids = 0
      
        for i in range(len(self.all_cars_list)):
            if int(self.all_cars_list[i][3]) == year:
                if self.all_cars_list[i][7] == True:
                    count_model_ids = count_model_ids + 1
      
        
        return count_model_ids;
      
       
        
        # Add logic for step 2 by looping through the cars_list variable and populating the 2 instance variables above.

In [8]:
carsdata = Cars('cars.json')
print(carsdata.count_hybrid(2009))
print(carsdata.find_max_horsepower(['2009 Audi A3 2.0 T AT','2009 Audi A3 2.0 T AT']))
print(carsdata.find_model_id('Audi'))
print(carsdata.get_all_cars_by_make())


4
(200, ['2009 Audi A3 2.0 T AT', '2009 Audi A3 2.0 T AT'])
(87, ['2009 Audi A3 3.2', '2009 Audi A3 2.0 T AT', '2009 Audi A3 2.0 T', '2009 Audi A3 2.0 T Quattro', '2009 Audi A3 2.0 T Quattro', '2009 Audi A5 3.2', '2009 Audi A5 3.2 AT', '2009 Audi Q7 4.2', '2009 Audi Q7 3.6', '2009 Audi A4 Sedan 2.0 T Quattro', '2009 Audi A4 Sedan 2.0 T Quattro AT', '2009 Audi A4 Sedan 3.2', '2012 Audi A3 2.0 T', '2011 Audi A3 2.0 T', '2010 Audi A3 2.0 T', '2012 Audi A4 Sedan 2.0 T Quattro', '2012 Audi A4 Sedan 2.0 T Quattro AT', '2011 Audi A4 Sedan 2.0 T Quattro', '2011 Audi A4 Sedan 2.0 T Quattro AT', '2010 Audi A4 Sedan 2.0 T Quattro AT', '2010 Audi A4 Sedan 2.0 T Quattro', '2012 Audi A4 Avant 2.0 T', '2011 Audi A4 Avant 2.0 T', '2010 Audi A4 Avant 2.0 T', '2011 Audi A5 Cabriolet 2.0 T Quattro', '2010 Audi A5 Cabriolet 2.0 T Quattro', '2012 Audi A5 Coupe 2.0 T', '2012 Audi A5 Coupe 2.0 T AT', '2011 Audi A5 Coupe 2.0 T AT', '2011 Audi A5 Coupe 2.0 T', '2010 Audi A5 Coupe 2.0 T', '2010 Audi A5 Coupe 3.

## Part 2 - Writing methods

Now that your Cars class has an *$__init__$* method and instance variables, you can instantiate objects by calling that class. The next step is creating methods that will operate on that object (and do something useful).

### 🎯 Create 4 methods to accomplish tasks 1 - 4 below:
You will use these methods in part 3 to answer questions.

### 1. Find the number of **model ID**'s for a given make. 
* Call this method "*find_model_id*"
* **Argument**: Make (string)
* **Return**s (2 values): Both the count and list of model IDs for a given Make.
* **Example method call**: num_ids, chevy_model_ids = carsdata.find_model_id("Chevrolet")

### 2. Retrieve maximum horsepower for a list of model ids.
* Call this method "*find_max_horsepower*"
* **Argument**: Model_ids (list)
* **Return** (1 value): tuple of max horsepower and model_id for a list of model_ids.
* **Example method call**: chevy_hp, chevy_hp_models = carsdata.find_max_horsepower(chevy_model_ids)

### 3. Calculate the highest city-mpg to highway-mpg ratio for given Make.
* Call this method "*find_mpgratio*"
* **Argument**: Make (string)
* **Return** (1 value): tuple of: Highway_mpg-to-city_mpg ratio and model ID.
* **Example method call**: chevy_mpg_ratio, chevy_mpg_models = carsdata.find_mpgratio("Chevrolet")
* **Note**: make sure you convert both mpg (highway and city) to float *before* calculating.

### 4. Calculate the number of hybrid model ids for a given year.
* Call this method "*count_hybrid*"
* **Argument**: Year (integer)
* **Return** (1 value): Count of hybrid model_ids
* **Example method call**: count_hybrid_2009 = carsdata.count_hybrid(2009)
* **Note**: The Cars data has limited years.


### Important notes
* Add these methods to your class in the code cell from part 1.
* For the purposes of this assignment, every method has the *self* parameter.
* Make sure that your methods account for ties.
* Do not be alarmed when you see text prefixed by 'u'. This stands for unicode, which is the default string type in Python3. Strings meant to be human-readable should always be in unicode.
* 🎯 After creating the class, first test it with the sample dataset (instantiate an object with 'cars_truncated.json'). 
* 🎯 If everything works, then test it on the full dataset.

## Part 3 - Answering questions

🎯 Use your class and methods to answer the questions below.

### 1. How many Audi *model_ids* are in this dataset, and what are the models?

In [9]:
num_ids, audi_model_ids = carsdata.find_model_id('Audi')
print(num_ids, '\n', audi_model_ids)

87 
 ['2009 Audi A3 3.2', '2009 Audi A3 2.0 T AT', '2009 Audi A3 2.0 T', '2009 Audi A3 2.0 T Quattro', '2009 Audi A3 2.0 T Quattro', '2009 Audi A5 3.2', '2009 Audi A5 3.2 AT', '2009 Audi Q7 4.2', '2009 Audi Q7 3.6', '2009 Audi A4 Sedan 2.0 T Quattro', '2009 Audi A4 Sedan 2.0 T Quattro AT', '2009 Audi A4 Sedan 3.2', '2012 Audi A3 2.0 T', '2011 Audi A3 2.0 T', '2010 Audi A3 2.0 T', '2012 Audi A4 Sedan 2.0 T Quattro', '2012 Audi A4 Sedan 2.0 T Quattro AT', '2011 Audi A4 Sedan 2.0 T Quattro', '2011 Audi A4 Sedan 2.0 T Quattro AT', '2010 Audi A4 Sedan 2.0 T Quattro AT', '2010 Audi A4 Sedan 2.0 T Quattro', '2012 Audi A4 Avant 2.0 T', '2011 Audi A4 Avant 2.0 T', '2010 Audi A4 Avant 2.0 T', '2011 Audi A5 Cabriolet 2.0 T Quattro', '2010 Audi A5 Cabriolet 2.0 T Quattro', '2012 Audi A5 Coupe 2.0 T', '2012 Audi A5 Coupe 2.0 T AT', '2011 Audi A5 Coupe 2.0 T AT', '2011 Audi A5 Coupe 2.0 T', '2010 Audi A5 Coupe 2.0 T', '2010 Audi A5 Coupe 3.2', '2010 Audi A5 Coupe 2.0 T AT', '2012 Audi A6 Sedan 3.0 T

### 2. Which Audi model has the most horsepower? Use your second method to find this answer.

In [10]:
(audi_maxhp, audi_maxhp_id) = carsdata.find_max_horsepower(audi_model_ids)
print(audi_maxhp_id)

['2011 Audi R8 Coupe 5.2', '2012 Audi R8 Coupe 5.2', '2010 Audi R8 Coupe 5.2', '2011 Audi R8 Spyder 5.2', '2012 Audi R8 Spyder 5.2']


### 3. Find the model with the highest city-to-highway mpg for Chevrolet. Print the ratio and model(s) that have this ratio.

In [11]:
(chevy_maxchratio, chevy_maxchratio_ids) = carsdata.find_mpgratio('Chevrolet')
print(chevy_maxchratio, '\n', chevy_maxchratio_ids)

1.0 
 ['2009 Chevrolet Silverado 1500 Hybrid 1HY 4WD', '2009 Chevrolet Silverado 1500 Hybrid 2HY 4WD']


### 4. How many hybrid cars were produced in 2009? How many in 2011?

In [12]:
hybrid_2009 = carsdata.count_hybrid(2009)
hybrid_2011 = carsdata.count_hybrid(2011)
 
print("2009:", hybrid_2009, "\n", "2011:", hybrid_2011)

2009: 4 
 2011: 26


#### Sources

<sup>[1]</sup> OOP, http://openbookproject.net/thinkcs/python/english3e/classes_and_objects_I.html<br/>
<sup>[2]</sup> Why use classes, https://stackoverflow.com/questions/33072570/when-should-i-be-using-classes-in-python<br/>
<sup>[3]</sup> Why use classes, https://dbader.org/blog/6-things-youre-missing-out-on-by-never-using-classes-in-your-python-code<br/>
<sup>[4]</sup> OOP, https://realpython.com/python3-object-oriented-programming/<br/>
<sup>[5]</sup> Class and Static methods, https://realpython.com/instance-class-and-static-methods-demystified/<br/>
<sup>[6]</sup> Self and __init__, https://micropyramid.com/blog/understand-self-and-__init__-method-in-python-class/<br/>
<sup>[7]</sup> Self and __init__, https://stackoverflow.com/questions/625083/python-init-and-self-what-do-they-do